### ISSUE TITLE
Non-deterministic behaviour of MeshRasterizer

### ISSUE THREAD

**Issue by nh236** [facebookresearch/pytorch3d#1519]
Hi,

first of all, thanks for all the great work!

## 🐛 Bugs / Unexpected behaviors
The output of MeshRasterizer seems to be non-deterministic. Calling the forward function with the same input multiple times produces (slightly) different outputs.

## Instructions To Reproduce the Issue:
The following code is adapted from the [rendering tutorial](https://github.com/facebookresearch/pytorch3d/blob/main/docs/tutorials/camera_position_optimization_with_differentiable_rendering.ipynb).
(The value of `faces_per_pixel` in the RasterizationSettings was changed from `100` to `10`. With `faces_per_pixel=100`, the output seemed to be deterministic. In my actual use case, however, I needed to reduce the number of `faces_per_pixel` due to higher resolution and GPU memory limit, which is why `faces_per_pixel` cannot be increased.)

I followed [pytorch instructions](https://pytorch.org/docs/stable/notes/randomness.html) to make the pytorch code as deterministic as possible.

The sample mesh is taken from [here](https://dl.fbaipublicfiles.com/pytorch3d/data/teapot/teapot.obj).

I'm using torch 1.13.1+cu116 and pytorch3d 0.7.2

```
import os
import random

import numpy as np
import torch

# determinsim: https://pytorch.org/docs/stable/notes/randomness.html
os.environ['CUBLAS_WORKSPACE_CONFIG'] = ":16:8"
random.seed(0)
np.random.seed(0)
torch.manual_seed(0)
torch.cuda.manual_seed_all(0)
torch.use_deterministic_algorithms(True)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

# io utils
from pytorch3d.io import load_obj

# datastructures
from pytorch3d.structures import Meshes

# rendering components
from pytorch3d.renderer import (
    FoVPerspectiveCameras, look_at_view_transform,
    RasterizationSettings, MeshRenderer, MeshRasterizer, BlendParams,
    SoftSilhouetteShader, TexturesVertex
)

def main():
    # Code from https://github.com/facebookresearch/pytorch3d/blob/main/docs/tutorials/camera_position_optimization_with_differentiable_rendering.ipynb

    # Set the cuda device
    if torch.cuda.is_available():
        device = torch.device("cuda:0")
        torch.cuda.set_device(device)
    else:
        device = torch.device("cpu")

    # Load the obj and ignore the textures and materials.
    verts, faces_idx, _ = load_obj("./data/teapot.obj")
    faces = faces_idx.verts_idx

    # Initialize each vertex to be white in color.
    verts_rgb = torch.ones_like(verts)[None]  # (1, V, 3)
    textures = TexturesVertex(verts_features=verts_rgb.to(device))

    # Create a Meshes object for the teapot. Here we have only one mesh in the batch.
    teapot_mesh = Meshes(
        verts=[verts.to(device)],
        faces=[faces.to(device)],
        textures=textures
    )

    # Initialize a perspective camera.
    cameras = FoVPerspectiveCameras(device=device)

    # To blend the 100 faces we set a few parameters which control the opacity and the sharpness of
    # edges. Refer to blending.py for more details.
    blend_params = BlendParams(sigma=1e-4, gamma=1e-4)

    # Define the settings for rasterization and shading. Here we set the output image to be of size
    # 256x256. To form the blended image we use 100 faces for each pixel. We also set bin_size and max_faces_per_bin to None which ensure that
    # the faster coarse-to-fine rasterization method is used. Refer to rasterize_meshes.py for
    # explanations of these parameters. Refer to docs/notes/renderer.md for an explanation of
    # the difference between naive and coarse-to-fine rasterization.
    raster_settings = RasterizationSettings(
        image_size=256,
        blur_radius=np.log(1. / 1e-4 - 1.) * blend_params.sigma,
        faces_per_pixel=10
    )

    # Create a silhouette mesh renderer by composing a rasterizer and a shader.
    silhouette_renderer = MeshRenderer(
        rasterizer=MeshRasterizer(
            cameras=cameras,
            raster_settings=raster_settings
        ),
        shader=SoftSilhouetteShader(blend_params=blend_params)
    )

    # Select the viewpoint using spherical angles
    distance = 3  # distance from camera to the object
    elevation = 50.0  # angle of elevation in degrees
    azimuth = 0.0  # No rotation so the camera is positioned on the +Z axis.

    # Get the position of the camera based on the spherical angles
    R, T = look_at_view_transform(distance, elevation, azimuth, device=device)

    # Render the teapot providing the values of R and T.
    # silhouette1 = silhouette_renderer(meshes_world=teapot_mesh, R=R, T=T)
    fragments1 = silhouette_renderer.rasterizer(teapot_mesh, R=R, T=T)
    # images1 = silhouette_renderer.shader(fragments1, teapot_mesh, R=R, T=T)

    # silhouette2 = silhouette_renderer(meshes_world=teapot_mesh, R=R, T=T)
    fragments2 = silhouette_renderer.rasterizer(teapot_mesh, R=R, T=T)
    # images2 = silhouette_renderer.shader(fragments2, teapot_mesh, R=R, T=T)

    diff_p2f = torch.where((fragments1.pix_to_face - fragments2.pix_to_face) != 0.)

    face_ids_1 = fragments1.pix_to_face[diff_p2f]
    dists1 = fragments1.dists[diff_p2f]
    face_ids_2 = fragments2.pix_to_face[diff_p2f]
    dists2 = fragments2.dists[diff_p2f]

    print('face inds 1: {}'.format(face_ids_1.detach().cpu().numpy().squeeze()))
    print('face inds 2: {}'.format(face_ids_2.detach().cpu().numpy().squeeze()))
    print('dists 1: {}'.format(dists1.detach().cpu().numpy().squeeze()))
    print('dists 2: {}'.format(dists2.detach().cpu().numpy().squeeze()))

if __name__=='__main__':
    main()
```

Sample output, showing the indices of faces that are different, with corresponding distances:
```
face inds 1: [1912 1912 1912 1912 1889 1996  583  589  602  602  591  193  992  601
  277  990  626 1668  626 1668  660  651  651  649  651  625  650 2282
 2282 2258 1590 1590]
face inds 2: [1899 1899 1899 1910 2242 2243  851  602  589  589  589  997 1001  614
 2393  999  613 2435  624 2372  651  649  649  651  660  627  648  994
  994  876 1210 1210]
dists 1: [6.9428334e-04 1.1204268e-04 8.7540138e-06 1.2976732e-04 1.8824830e-04
 1.8824830e-04 5.5223744e-04 2.3745060e-04 1.6553263e-04 1.6763351e-04
 6.6642574e-04 8.4119686e-04 2.0514675e-04 1.7616730e-04 5.5017008e-04
 9.2627351e-06 2.8225116e-04 6.8947492e-04 3.1065251e-04 5.9729262e-04
 2.9225138e-04 6.3290045e-04 3.4247333e-04 1.7031487e-04 1.1295908e-04
 6.1635231e-04 7.5895106e-04 7.3865178e-04 7.9468114e-04 4.0886813e-04
 6.3299114e-04 7.7149406e-04]
dists 2: [6.7892531e-04 1.1442895e-04 3.4412478e-05 3.8469597e-04 8.5381820e-05
 8.5381820e-05 8.3158637e-04 3.6559085e-04 1.4778810e-04 2.8002617e-04
 5.4016057e-04 6.5047032e-05 4.1431360e-04 2.9302624e-04 8.2154974e-04
 3.1736074e-06 2.7200187e-04 6.8947492e-04 3.9405926e-04 5.9729262e-04
 3.1464006e-04 6.7971810e-04 3.6398129e-04 1.7029083e-04 1.4919294e-04
 6.1635231e-04 7.5895106e-04 5.4196769e-04 5.3275301e-04 2.2641502e-04
 6.8788638e-04 7.3166966e-04]
```

Do you have any idea where this behaviour is coming from (and how to make the rasterizer truly deterministic)?

In an [earlier issue](https://github.com/facebookresearch/pytorch3d/issues/659), a similar problem was attributed to multiple faces having exactly the same distances, which was supposedly [resolved](https://github.com/facebookresearch/pytorch3d/commit/860b742a02e62ec85f48929268349916ca4ce8a5). This doesn't seem to be the problem here, since the distances are not the same (`dists_1` vs `dists_2`).

Thank you for your help!

Best,
Nikolas

**Comment by bottler** [comment]
I don't think the cuda parts of rasterizer forward pass are nondeterministic. It might be interesting to retry setting `bin_size=0` in the RasterizationSettings, which causes a simpler (and often slower) version of the rasterizer to be used.

**Comment by nh236** [comment]
Thank you for your answer.

Setting bin_size=0 does lead to deterministic results in the _forward_ pass. However, when I use this in an optimization loop (fitting a mesh to a silhouette), the loss is still not deterministic, e.g. in the camera position optimization tutorial (see below for a slightly changed version saving the losses to files and comparing the current loss to the saved loss when the code is run a second time).

```
import os
import random

import numpy as np
import torch

# determinsim: https://pytorch.org/docs/stable/notes/randomness.html
os.environ['CUBLAS_WORKSPACE_CONFIG'] = ":16:8"
random.seed(0)
np.random.seed(0)
torch.manual_seed(0)
torch.cuda.manual_seed_all(0)
torch.use_deterministic_algorithms(True)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

import torch.nn as nn
import matplotlib.pyplot as plt

# io utils
from pytorch3d.io import load_obj

# datastructures
from pytorch3d.structures import Meshes


# rendering components
from pytorch3d.renderer import (
    FoVPerspectiveCameras, look_at_view_transform, look_at_rotation, 
    RasterizationSettings, MeshRenderer, MeshRasterizer, BlendParams,
    SoftSilhouetteShader, HardPhongShader, PointLights, TexturesVertex,
)

def temp_loss_check(save_folder, name, value, cnt):
    from os.path import join, exists

    fn = join(save_folder, '{:03d}_{}.npy'.format(cnt, name))

    if not exists(fn):
        np.save(fn, value)
    else:
        oldval = np.load(fn)

        if np.max(oldval - value) > 0:
            print('iter {} {}: max diff: {}'.format(cnt, name, np.max(oldval - value)))

# Set the cuda device 
if torch.cuda.is_available():
    device = torch.device("cuda:0")
    torch.cuda.set_device(device)
else:
    device = torch.device("cpu")

# Load the obj and ignore the textures and materials.
verts, faces_idx, _ = load_obj("./data/teapot.obj")
faces = faces_idx.verts_idx

# Initialize each vertex to be white in color.
verts_rgb = torch.ones_like(verts)[None]  # (1, V, 3)
textures = TexturesVertex(verts_features=verts_rgb.to(device))

# Create a Meshes object for the teapot. Here we have only one mesh in the batch.
teapot_mesh = Meshes(
    verts=[verts.to(device)],   
    faces=[faces.to(device)], 
    textures=textures
)

# Initialize a perspective camera.
cameras = FoVPerspectiveCameras(device=device)

# To blend the 100 faces we set a few parameters which control the opacity and the sharpness of 
# edges. Refer to blending.py for more details. 
blend_params = BlendParams(sigma=1e-4, gamma=1e-4)

# Define the settings for rasterization and shading. Here we set the output image to be of size
# 256x256. To form the blended image we use 100 faces for each pixel. We also set bin_size and max_faces_per_bin to None which ensure that 
# the faster coarse-to-fine rasterization method is used. Refer to rasterize_meshes.py for 
# explanations of these parameters. Refer to docs/notes/renderer.md for an explanation of 
# the difference between naive and coarse-to-fine rasterization. 
raster_settings = RasterizationSettings(
    image_size=256, 
    blur_radius=np.log(1. / 1e-4 - 1.) * blend_params.sigma, 
    faces_per_pixel=100,
    bin_size=0
)

# Create a silhouette mesh renderer by composing a rasterizer and a shader. 
silhouette_renderer = MeshRenderer(
    rasterizer=MeshRasterizer(
        cameras=cameras, 
        raster_settings=raster_settings
    ),
    shader=SoftSilhouetteShader(blend_params=blend_params)
)


# We will also create a Phong renderer. This is simpler and only needs to render one face per pixel.
raster_settings = RasterizationSettings(
    image_size=256, 
    blur_radius=0.0, 
    faces_per_pixel=1,
    bin_size=0
)
# We can add a point light in front of the object. 
lights = PointLights(device=device, location=((2.0, 2.0, -2.0),))
phong_renderer = MeshRenderer(
    rasterizer=MeshRasterizer(
        cameras=cameras, 
        raster_settings=raster_settings
    ),
    shader=HardPhongShader(device=device, cameras=cameras, lights=lights)
)


# Select the viewpoint using spherical angles  
distance = 3   # distance from camera to the object
elevation = 50.0   # angle of elevation in degrees
azimuth = 0.0  # No rotation so the camera is positioned on the +Z axis. 

# Get the position of the camera based on the spherical angles
R, T = look_at_view_transform(distance, elevation, azimuth, device=device)

# Render the teapot providing the values of R and T. 
silhouette = silhouette_renderer(meshes_world=teapot_mesh, R=R, T=T)
image_ref = phong_renderer(meshes_world=teapot_mesh, R=R, T=T)

silhouette = silhouette.cpu().numpy()
image_ref = image_ref.cpu().numpy()

class Model(nn.Module):
    def __init__(self, meshes, renderer, image_ref):
        super().__init__()
        self.meshes = meshes
        self.device = meshes.device
        self.renderer = renderer
        
        # Get the silhouette of the reference RGB image by finding all non-white pixel values. 
        image_ref = torch.from_numpy((image_ref[..., :3].max(-1) != 1).astype(np.float32))
        self.register_buffer('image_ref', image_ref)
        
        # Create an optimizable parameter for the x, y, z position of the camera. 
        self.camera_position = nn.Parameter(
            torch.from_numpy(np.array([3.0,  6.9, +2.5], dtype=np.float32)).to(meshes.device))

    def forward(self):
        
        # Render the image using the updated camera position. Based on the new position of the 
        # camera we calculate the rotation and translation matrices
        R = look_at_rotation(self.camera_position[None, :], device=self.device)  # (1, 3, 3)
        T = -torch.bmm(R.transpose(1, 2), self.camera_position[None, :, None])[:, :, 0]   # (1, 3)
        
        image = self.renderer(meshes_world=self.meshes.clone(), R=R, T=T)
        
        # Calculate the silhouette loss
        loss = torch.sum((image[..., 3] - self.image_ref) ** 2)
        return loss, image
  

# Initialize a model using the renderer, mesh and reference image
model = Model(meshes=teapot_mesh, renderer=silhouette_renderer, image_ref=image_ref).to(device)

# Create an optimizer. Here we are using Adam and we pass in the parameters of the model
optimizer = torch.optim.Adam(model.parameters(), lr=0.05)

# save temporary losses to check for non-determinism
folder = './data/losses'
if not os.path.exists(folder):
    os.makedirs(folder)

for i in range(200):
    optimizer.zero_grad()
    loss, _ = model()
    loss.backward()
    optimizer.step()

    print('Optimizing (loss {})'.format(loss.data))
    temp_loss_check(save_folder=folder, name='loss', value=loss.data.detach().cpu().numpy().squeeze(), cnt=i)

    if loss.item() < 200:
        break
```
It would really help me to have a fully deterministic pipeline, so I appreciate any help!

Thank you in advance!

Best,
Nikolas

**Comment by bottler** [comment]
The backward pass is known to not be deterministic as it [relies *solely* on atomic operations](https://github.com/facebookresearch/pytorch3d/blob/v0.7.4/pytorch3d/csrc/rasterize_meshes/rasterize_meshes.cu#L546-L561) to allow multiple threads to increment gradients of the same vertex position. In the latest version of PyTorch3D we now warn or fail for this if you try to use torch.use_deterministic_algorithms. (I recently did a sweep through the CUDA code to add these warnings. It sounds like I missed something in the forward pass.)

In fact the backward pass's code is much simpler than the forward pass. It would be relatively easy to rewrite the backward pass in PyTorch in Python, and I think the performance would not be hugely worse but the memory usage would be a bit bigger. Then the deterministic version of index_add_ would solve the problem.

**Comment by nh236** [comment]
OK, thanks a lot for your help!
